src/dune_pkg/rev_store.ml

```diff
@@ -3,42 +3,60 @@ open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+type t =
+  { dir : Path.t
+  ; lock : Flock.t
+  }
+
 type rev = Rev of string
 
-let equal { dir } t = Path.equal dir t.dir
+let rec lock_limited flock lock timeout =
+  let sleep_duration = 0.1 in
+  match Flock.lock_non_block flock lock with
+  | Ok `Failure as v ->
+    if timeout > 0.0
+    then (
+      let timeout = timeout -. sleep_duration in
```

Comment:
You should make sure `timeout` is non-negative here. E.g. `let timeout = max 0. timeout`.